4 week

duswnsxnxn opened the pull request:
# 4Week_김연준.md

## Title: [4Week] 김연준

### 필수 미션 요구사항 분석 & 체크리스트
- [x] 네이버클라우드플랫폼을 통한 배포, 도메인, HTTPS 까지 적용
- [x] 내가 받은 호감리스트(/usr/likeablePerson/toList)에서 성별 필터링기능 구현
### 선택 미션 요구사항 분석 & 체크리스트
- [x] 젠킨스를 통해서 리포지터리의 main 브랜치에 커밋 이벤트가 발생하면 자동으로 배포가 진행되도록
- [x] 내가 받은 호감리스트(/usr/likeablePerson/toList)에서 호감사유 필터링기능 구현
- [x] 내가 받은 호감리스트(/usr/likeablePerson/toList)에서 정렬기능
---

-  젠킨스를 통해서 리포지터리의 main 브랜치에 커밋 이벤트가 발생하면 자동으로 배포가 진행되도록 (미구현)
-  내가 받은 호감리스트(/usr/likeablePerson/toList)에서 호감사유 필터링기능 구현 (구현)
-  내가 받은 호감리스트(/usr/likeablePerson/toList)에서 정렬기능 (구현)
-  네이버클라우드플랫폼을 통한 배포, 도메인, HTTPS 까지 적용 (구현)
-  내가 받은 호감리스트(/usr/likeablePerson/toList)에서 성별 필터링기능 구현 (구현)
### 4주차 미션 요약

---

**[접근 방법]**



- 성별, 호감사유 필터링 기능
  - 컨트롤러에서 gender, attractiveTypeCode라는 이름의 파라미터를 받는다.
  - 컨트롤러에서 받은 파라미터를 querydsl레파지토리까지 전달
  - 처음 toList페이지를 들어갈 때 조건 파라미터들이 null값으로 전달되는 것을 고려하여 service단에서 처리하지 않고 querydsl로 처리함
  - BooleanExpression은 null일 경우 null을 반환하고 where절에서 자동으로 사라지는 특성이 있으므로 유연하게 동적쿼리에 대응
  - null이 아니라면 BooleanExpression는 원하는 필터링 조건을 반환하고 where절에 주입
- 정렬 기능
  - orderby절에는 표현식으로 OrderSpecifier클래스를 받으므로 이를 반환하는 sortByField 메소드를 구성
  - 클라이언트로부터 sortCode라는 1~6의 정렬 옵션 중 하나를 파라미터로 받아 switch문으로 각각의 옵션 처리 
  - 5, 6번 옵션은 다중 옵션이므로 sortByField메소드에서 OrderSpecifier[]와 같이 옵션을 배열에 담아 반환 
  - 정렬옵션을 선택하지 않을 때 기본으로 실행되는 옵션은 최신날짜순이므로 옵션파라미터가 null이거나 공백이면 기본옵션을 실행하게 처리
  - OrderSpecifier의 Expression 매개변수로 실제 엔티티에 존재하는 필드(데이터베이스에 등록된 필드)만 인식한다.
  - 인기 순으로 정렬하려면 호감을 표시한 인스타 멤버의 총 like 갯수로 판단해야하므로 InstaMemberBase에 likes필드를 추가
- 도메인, HTTPS적용
  - iwantmyname에서 도메인을 구입하고 dnszi에서 관리를 하여 도메인이 네이버 클라우드서버에서 받은 공인ip를 가르키도록 A레코드 설정
  - 접속용 포트인 443, 80 관리자 포트인 81번을 npm을 도입하여 전부 몰아준다.
  - npm에서 SSL인증서 발급받고 https적용

**[특이사항]**

아쉬웠던 점
- 인기 순 정렬에 likes를 기준으로 판단 할 다른 방법이 생각나지 않아 OrderSpecifier규격에 맞추려고 likes필드를 추가시켜 테이블 구조를 바꾼 것
- 배포 과정에 대해 전반적인 이해가 부족한 상태여서 다시 공부할 필요성을 느낌
- 젠킨스 자동 배포를 하지 못한 점

궁금했던 점
- 서비스단과 DB단의 정렬 처리 중 상황에 따라서 어느 방법이 좋은지 구체적으로 알고싶습니다.


**[Refactoring]**

duswnsxnxn committed:
feat: 성별 필터링기능 구현

duswnsxnxn committed:
feat: 호감사유 필터링기능 구현

duswnsxnxn committed:
feat: 젠킨스 제외 필수, 선택 미션 완료

duswnsxnxn committed:
test: 성별 필터링 테스트 추가

duswnsxnxn committed:
docs: 4Week리드미 추가

duswnsxnxn committed:
docs: 4Week리드미 추가

duswnsxnxn committed:
test: 테스트 수정

duswnsxnxn committed:
Merge branch '4Week' of https://github.com/duswnsxnxn/Mission_KimYeonJoon into 4Week

duswnsxnxn committed:
docs: 4주차 레코드 수정

dw2023 commented on src/main/java/com/ll/gramgram/boundedContext/instaMember/entity/InstaMemberBase.java:
```diff
@@ -16,6 +16,8 @@
 public abstract class InstaMemberBase extends BaseEntity {
     @Setter
     String gender;
+    //OrderSpecifier의 Expression으로 이용하여 정렬 조건에 만족시키기 위해 컬럼 추가
+    long likes;
```
성별/호감사유별 호감수를 다 더해주는 컬럼 추가하여 로직이 깔끔하고 가독성이 좋은 것 같습니다.

dw2023 reviewed (commented):
record에 전체적인 구현과정을 명료하게 남겨주셔서 이해하기 편했습니다!
QueryDSL을 사용해서 코드를 잘 작성해주셔서 인상 깊었고 덕분에 코드리뷰하면서 많은 공부가 되었습니다. 
좋은 코드 감사합니다!
다만 제가 아직 QueryDSL에 대한 이해가 부족하여 더 깊이있는 피드백을 드리지 못한 것 같아 개인적으로 아쉽네요

테스트 코드도 거의 다 구현하신 것 같은데 리팩토링하면서 마무리하시면 좋을 것 같습니다.

수고하셨습니다!

JunTaeINC reviewed (commented):
쿼리 DSL을 잘 사용하셨네요. 아직 배워야 할게 너무 많네요... 이번 미션도 구현하느라 고생많으셨습니다

duswnsxnxn commented on src/main/java/com/ll/gramgram/boundedContext/likeablePerson/controller/LikeablePersonController.java:
```diff
@@ -120,16 +120,29 @@ public String modify(@PathVariable Long id, @Valid ModifyForm modifyForm) {
         return rq.redirectWithMsg("/usr/likeablePerson/list", rsData);
     }
 
+    @AllArgsConstructor
+    @Getter
+    public static class ToListCondition {
+        private final String gender = null;
+        private final int attractiveTypeCode = 0;
+        private final int sortCode = 0;
+    }
```
아 저거 처음에 오브젝트로 파라미터 받으려다가 요청body에 담는게 아니라 쿼리스트링으로 보내는거라서 안돼더라구여 ㅠ 그래서 지우려고 했는데 깜빡했어요.
리펙토링 작업 시 지워야겠네요 감사합니다.

dw2023 replied:
아 그러셨군요 알려주셔서 감사합니다!

duswnsxnxn reviewed (commented):
깜빡하고 안지운 코드 알려주셔서 감사합니다.
덕분에 리펙토링 시 참고가 많이 될 것 같아요. ㅎㅎ

duswnsxnxn commented:
> 쿼리 DSL을 잘 사용하셨네요. 아직 배워야 할게 너무 많네요... 이번 미션도 구현하느라 고생많으셨습니다

미션에만 열중해서 코테연습을 안한지 너무 오래댔네여 ㅠㅠ 망했어요

duswnsxnxn committed:
refactor: 리펙토링
- ToListCondition 내부 클래스로 파라미터 바인딩으로 수정
- attractiveTypeCode, sortCode 타입 int형으로 변경
- LikeablePersonRepositoryImpl의 findQslByGenderAndAttractiveTypeCode메소드 예외처리 최적화

duswnsxnxn committed:
test: 성별 필터링 테스트 수정

jhs512 commented on src/main/java/com/ll/gramgram/boundedContext/likeablePerson/repository/LikeablePersonRepositoryImpl.java:
```diff
@@ -26,4 +32,63 @@ public Optional<LikeablePerson> findQslByFromInstaMemberIdAndToInstaMember_usern
+        if(gender.isEmpty()) {
+            return null;
+        }
+        return likeablePerson.fromInstaMember.gender.eq(gender);
+    }
+
+    private BooleanExpression eqAttractiveTypeCode(InstaMember instaMember, int attractiveTypeCode) {
+        if(attractiveTypeCode == 0) {
+            return null;
+        }
+        return likeablePerson.attractiveTypeCode.eq(attractiveTypeCode);
+    }
+
+    private OrderSpecifier[] sortByField(int sortCode) {
+
+        List<OrderSpecifier> orderSpecifiers = new ArrayList<>();
+
+
+        switch (sortCode) {
+            case 1 -> orderSpecifiers.add(new OrderSpecifier<>(Order.DESC, likeablePerson.createDate));
+            case 2 -> orderSpecifiers.add(new OrderSpecifier<>(Order.ASC, likeablePerson.createDate));
+            case 3 -> {
+                orderSpecifiers.add(new OrderSpecifier<>(Order.DESC, likeablePerson.fromInstaMember.likes));
+                orderSpecifiers.add(new OrderSpecifier<>(Order.DESC, likeablePerson.createDate));
+            }
+            case 4 -> {
+                orderSpecifiers.add(new OrderSpecifier<>(Order.ASC, likeablePerson.fromInstaMember.likes));
+                orderSpecifiers.add(new OrderSpecifier<>(Order.DESC, likeablePerson.createDate));
+            }
+            case 5 -> {
+                orderSpecifiers.add(new OrderSpecifier<>(Order.DESC, likeablePerson.fromInstaMember.gender));
+                orderSpecifiers.add(new OrderSpecifier<>(Order.DESC, likeablePerson.createDate));
+            }
+            case 6 -> {
+                orderSpecifiers.add(new OrderSpecifier<>(Order.ASC, likeablePerson.attractiveTypeCode));
+                orderSpecifiers.add(new OrderSpecifier<>(Order.DESC, likeablePerson.createDate));
+            }
+        }
+        //다중 정렬 조건을 고려하여 배열로 반환
+        return orderSpecifiers.toArray(new OrderSpecifier[orderSpecifiers.size()]);
```
정렬을 할 때 정렬의 기준이 되는 필드에 인덱스가 걸려있다면 좋습니다. 그런면에서 봤을 때 잘 하신 선택입니다. 정렬만을 위해서 칼럼을 만드는게 낭비처럼 느껴질 수 있습니다. 하지만 이 방식은 많이 사용하는 방식입니다.
